src/api/app/services/monitor_controller_service/building_information_updater.rb

```diff
@@ -0,0 +1,50 @@
+module MonitorControllerService
+  class BuildingInformationUpdater
+    attr_accessor :workers
+
+    def initialize
+      @workers = {}
+    end
+
+    def call
+      initialize_workers
+      update_workers
+      self
+    end
+
+    private
+
+    def worker_status
+      @worker_status ||= Xmlhash.parse(WorkerStatus.hidden.to_xml)
+    end
+
+    def initialize_workers
+      @workers = Hash[worker_status.elements('idle').collect { |b| [worker_id(b), {}] }]
+    end
+
+    def update_workers
+      @workers.merge!(Hash[@worker_status.elements('building').collect { |b| [worker_id(b), workers_hash(b, calculate_delta(b['starttime'].to_i))] }])
+    end
+
+    def workers_hash(b, delta)
+      Hash[['project', 'repository', 'package', 'arch', 'starttime'].collect { |k| [k, b[k]] }].merge('delta' => delta)
+    end
+
+    def calculate_delta(starttime)
+      delta = (Time.now - Time.at(starttime)).round
+      delta = 5 if delta < 5
+      delta = max_time if delta > max_time
+      delta = (100 * Math.sin(Math.acos(1 - (Float(delta) / max_time)))).round
+      delta = 100 if delta > 100
+      delta
+    end
+
+    def worker_id(b)
+      b['workerid'].gsub(%r{[:./]}, '_')
+    end
+
+    def max_time
+      4 * 3600
```

Comment:
What about defining a constant?

Reply:
because it get evaluated when the code is loaded which isn't really what we want for each calculation that we have.

Reply:
```
vpereira@redstar:~> cat c.rb
FOO = 2/0

puts "bar"
vpereira@redstar:~> ruby c.rb 
Traceback (most recent call last):
	1: from c.rb:1:in `<main>'
c.rb:1:in `/': divided by 0 (ZeroDivisionError)
vpereira@redstar:~> 
```

Reply:
I have another problem. Those are magic numbers. What are those? Seconds I guess. Minutes? And why not use `14400` in 36 then? :-)

Reply:
I went with `4.hours.to_i`